crates/libs/windows/src/core/pcwstr.rs

```diff
@@ -45,7 +45,7 @@ unsafe impl Abi for PCWSTR {
 #[cfg(feature = "alloc")]
 impl<'a> IntoParam<'a, PCWSTR> for &str {
     fn into_param(self) -> Param<'a, PCWSTR> {
-        Param::Boxed(PCWSTR(heap_string(&self.encode_utf16().collect::<alloc::vec::Vec<u16>>())))
+        Param::Boxed(PCWSTR(unsafe { string_from_iter(self.encode_utf16(), self.len()) }))
```

Comment:
What if you just embrace the fact that you know `self` is a `&str` and pass in the actual length instead of over-allocating?  I don't know how expensive an over-allocation is relative to the recalculation.

```rust
string_from_iter(self.encode_utf16(), self.encode_utf16().count())
```

Reply:
We would need to iterate over it twice, which provides some memory savings at the cost of run-time computation.

The amount of memory we waste is 0% if the string consists of ASCII only.
The amount of memory we waste is 50% if the string consists of BMP code points (that are not ASCII) only.
The amount of memory we waste is 50% if the string consists of supplementary code points only (requires surrogate pairs for UTF-16 (2 chars) and 4 byte-sequences for UTF-8 (4 chars))

Given that [most strings are ASCII or contain mostly ASCII](utf8everywhere.org/), I believe it is a reasonable choice.

When considering memory waste we have to look at the intention the `IntoParam` trait provides: It is a convenience API.

In heavily resource constrained Windows environment such as [Nano server](https://docs.microsoft.com/en-us/windows-server/get-started/getting-started-with-nano-server), with a minimum requirement of [512 MB](https://docs.microsoft.com/en-us/windows-server/get-started/hardware-requirements), I'd believe heap allocation is frowned upon as is case in most embedded environments. I don't think this convenience API is something that is intended for Windows Nano environments and even if, given small strings the waste is still comparatively low.

Given the 2 GB memory requirement of Windows, the only way in which string allocation would be worry-some would be strings with length of 512 MB. Aside from the fact that it would be ludicrous to keep such huge strings in memory and not maintain/convert for yourself for performance reasons, these strings are also only allocated for the duration of the function call, they are immediately freed after.

I'd say given the use case, only a very low amount of PCs [having memory with less than 4 GB](https://store.steampowered.com/hwsurvey) and that most strings aren't occupying some ludicrous size, I'd say that over-allocating here is a reasonable choice to make. The mostly practical application of supplementary code points or BMP code-points is mostly [foreign languages](https://unicode-table.com/en/blocks/) that due to their amount of characters convey more entropy in their texts anyway and are hence smaller.

That being said, allocation the optimal length is still a desirable goal. Additionally, reading from the memory when decoding the unicode code points will likely enter the processor cache, which should make the the second iteration faster.  I'll rewrite something later that takes advantage of this.